pkg/lifecycle/service.go

```diff
@@ -211,49 +287,66 @@ func (s *Service) Wait(timeout time.Duration) error {
 	case <-gracefullyStopped:
 		return err
 	case <-time.After(timeout):
-		return ErrTimeout
+		return pipeline.ErrTimeout
 	}
 }
 
 // waitInternal blocks until all pipelines are stopped and returns an error if any of
 // the pipelines failed to stop gracefully.
 func (s *Service) waitInternal() error {
 	var errs []error
-	for _, pl := range s.instances {
-		err := pl.Wait()
+
+	s.m.Lock()
+	pipelines := s.runningPipelines
+	s.m.Unlock()
```

Comment:
Here we have a similar issue, although here we can't really hold the lock, since stopping a pipeline also requires the lock, and we are waiting for pipelines to stop. So here we would actually need to make a copy in advance so we can safely iterate over all the pipelines.

This would indeed be easier if we used the map from [this PR](https://github.com/ConduitIO/conduit-commons/pull/116), as we could simply do `s.runningPipelines.Copy()`.

Reply:
@lovromazgon https://github.com/ConduitIO/conduit/pull/1835/commits/3596fb82428f71d8d990aed272d2407fd2344157